src/modules/crypto/crypto_adaptors/common.rs

```diff
@@ -0,0 +1,339 @@
+            if $aes.ctx.as_mut().unwrap().aad_set == false {
+                if let Some(aad) = &$aes.aad {
+                    $aes.ctx.as_mut().unwrap().ctx.aad_update(aad).unwrap();
+                    $aes.ctx.as_mut().unwrap().aad_set = true;
+                }
+            }
+        }
+
+        // do real jobs.
+        // this Crypter::update returns a Result<usize, ErrorStack>, we simply ignore the detailed
+        // error information by unwrapping it.
+        // we also can't use the question mark operatior since the error codes are differently
+        // defined in RustyVault and underlying adaptor, such as rust-openssl.
+        let count = $aes.ctx.as_mut().unwrap().ctx.update(&$plaintext, &mut $ciphertext[..]).unwrap();
+
+        return Ok(count);
+    }
+}
+
+macro_rules! common_aes_encrypt_final {
+    ($aes: expr, $ciphertext: expr) => {
+        // Unlike encrypt_update() function, we don't do auto-initialization here.
+        if let None = $aes.ctx {
+            return Err(RvError::ErrCryptoCipherNotInited);
+        }
+
+        let count = $aes.ctx.as_mut().unwrap().ctx.finalize($ciphertext).unwrap();
+
+        if $aes.alg.1 == CipherMode::GCM {
+            // set tag for caller to obtain.
+            if let Some(_) = $aes.tag {
+                // tag should not be set before encrypt_final() is called.
+                return Err(RvError::ErrCryptoCipherAEADTagPresent);
+            }
+
+            // 16-byte long is enough for all types of AEAD cipher tag.
+            // TODO: this is for AES-128-GCM only.
+            let mut tag: Vec<u8> = vec![0; 16];
+            $aes.ctx.as_mut().unwrap().ctx.get_tag(&mut tag).unwrap();
+            $aes.tag = Some(tag);
+        }
+
+        return Ok(count);
+    }
+}
+
+macro_rules! common_aes_decrypt_update {
+    ($aes: expr, $ciphertext: expr, $plaintext: expr) => {
+        let cipher;
+
+        match $aes.alg {
+            (AESKeySize::AES128, CipherMode::CBC) => {
+                cipher = Cipher::aes_128_cbc();
+            }
+            (AESKeySize::AES128, CipherMode::GCM) => {
+                cipher = Cipher::aes_128_gcm();
+            }
+            _ => { return Err(RvError::ErrCryptoCipherOPNotSupported); }
+        }
+
+        if let None = $aes.ctx {
+            // init adaptor ctx if it's not inited.
+            let encrypter = Crypter::new(
+                cipher,
+                Mode::Decrypt,
+                &$aes.key,
+                Some(&$aes.iv)
+            ).unwrap();
+            let adaptor_ctx = AdaptorCTX { ctx: encrypter, tag_set: false, aad_set: false };
+
+            $aes.ctx = Some(adaptor_ctx);
+        }
+
+        // set additional authenticated data before doing real jobs.
+        if $aes.alg.1 == CipherMode::GCM {
+            if $aes.ctx.as_mut().unwrap().aad_set == false {
+                if let Some(aad) = &$aes.aad {
+                    $aes.ctx.as_mut().unwrap().ctx.aad_update(aad).unwrap();
```

Comment:
Same as above.

Reply:
Variable add is used later, so no need to change to `is_some`